firebase_admin/auth.py

```diff
@@ -94,10 +96,653 @@ def verify_id_token(id_token, app=None):
+
+        Returns:
+          bool: True if the user account is disabled, and False otherwise.
+        """
+        return bool(self._data.get('disabled'))
+
+    @property
+    def user_metadata(self):
+        """Returns additional metadata associated with this user.
+
+        Returns:
+          UserMetadata: A UserMetadata instance. Does not return None.
+        """
+        return UserMetadata(self._data)
+
+    @property
+    def provider_data(self):
+        """Returns a list of UserInfo instances.
+
+        Each object represents an identity from an identity provider that is linked to this user.
+
+        Returns:
+          list: A list of UserInfo objects, which may be empty.
+        """
+        providers = self._data.get('providerUserInfo', [])
+        return [_ProviderUserInfo(entry) for entry in providers]
+
+
+class UserMetadata(object):
+    """Contains additional metadata associated with a user account."""
+
+    def __init__(self, data):
+        if not isinstance(data, dict):
+            raise ValueError('Invalid data argument: {0}. Must be a dictionary.'.format(data))
+        self._data = data
+
+    @property
+    def creation_timestamp(self):
+        if 'createdAt' in self._data:
+            return int(self._data['createdAt'])
+        return None
+
+    @property
+    def last_sign_in_timestamp(self):
+        if 'lastLoginAt' in self._data:
+            return int(self._data['lastLoginAt'])
+        return None
+
+
+class _ProviderUserInfo(UserInfo):
+    """Contains metadata regarding how a user is known by a particular identity provider."""
+
+    def __init__(self, data):
+        super(_ProviderUserInfo, self).__init__()
+        if not isinstance(data, dict):
+            raise ValueError('Invalid data argument: {0}. Must be a dictionary.'.format(data))
+        if not data.get('rawId'):
+            raise ValueError('User ID must not be None or empty.')
+        self._data = data
+
+    @property
+    def uid(self):
+        return self._data.get('rawId')
+
+    @property
+    def display_name(self):
+        return self._data.get('displayName')
+
+    @property
+    def email(self):
+        return self._data.get('email')
+
+    @property
+    def phone_number(self):
+        return self._data.get('phoneNumber')
+
+    @property
+    def photo_url(self):
+        return self._data.get('photoUrl')
+
+    @property
+    def provider_id(self):
+        return self._data.get('providerId')
+
+
+class AuthError(Exception):
+    """Represents an Exception encountered while invoking the Firebase auth API."""
+
+    def __init__(self, code, message, error=None):
+        Exception.__init__(self, message)
+        self.code = code
+        self.detail = error
+
+
+class _Validator(object):
+    """A collectoin of data validation utilities.
+
+    Method provided in this class raise ValueErrors if any validations fail. Normal returns
+    signal success.
+    """
+
+    EMAIL_PATTERN = re.compile('^[^@]+@[^@]+$')
```

Comment:
Changed the validation as suggested. We only need minimal checking here. Backend does more comprehensive checking.